FranzKrekeler/vinecopulaslab/partnerselection/extremal.py

```diff
@@ -0,0 +1,99 @@
+from typing import List
+import itertools
+import logging
+import pandas as pd
+import numpy as np
+import scipy.special
+import scipy.linalg
+from vinecopulaslab.partnerselection.base import SelectionBase
+
+
+class ExtremalSelection(SelectionBase):
+    """
+    Class for partner selection based on "A multivariate linear rank test of independence
+    based on a multiparametric copula with cubic sections"
+    Mangold 2015
+    """
+
+    def __init__(self):
+        """Initialization
+        """
+        super().__init__()
+        self.corr_returns_top_n = None
+
+    def _partner_selection_approach(self, group):
+        """
+        Approach function Partner selection based on "A multivariate linear rank test of independence based on
+        a multiparametric copula with cubic sections"
+        for df.groupby("TARGET_STOCK").apply(...)
+        This has only been implemented for performance testing.
+        References:
+        https://www.researchgate.net/publication/309408947_A_multivariate_linear_rank_test_of_independence_based_on_a_multiparametric_copula_with_cubic_sections
+        https://pypi.org/project/Independence-test/
+        :param group: (group) The group of 50 most correlated stocks
+        :return: (List[str]) returns a list of highest correlated quadruple
+        """
+        logging.warning("Extremal approach is still under construction")
+        target_stock = group.name
+        partner_stocks = group.STOCK_PAIR.tolist()
+        stock_selection = [target_stock] + partner_stocks
+        # We create a subset of our ranked returns dataframe to increase lookup speed.
+        data_subset = self.ranked_returns[stock_selection].copy()
+        # We turn our partner stocks into numerical indices so we can use them directly for indexing
+        quadruples_combinations = self._prepare_combinations_of_partners(stock_selection)
+        # We can now use our list of possible quadruples as an index
+        quadruples_combinations_data = data_subset.values[:, quadruples_combinations]
+        # Now we can get closer to a vectorized calculation
+        # n is equal to the total number of returns d to the number of stocks
+        # we use lodash because we don't need the 19600 dimension
+        n, _, d = quadruples_combinations_data.shape
+        # Here the math from the Mangold 2015 paper begins
+        permut_mat = np.array(list(itertools.product([-1, 1], repeat=d)), dtype=np.int8)
+        sub_mat = permut_mat @ permut_mat.T
+        F = (d + sub_mat) / 2
```

Comment:
I'm not a fan of variables that are all CAPS unless they are env variables. Maybe you tried to match the notation in the paper? If not then perhaps a more descriptive name would help.